tests/legacy/unit/pluginhandler/test_environment.py

```diff
@@ -0,0 +1,86 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2023 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+from snapcraft_legacy.internal.pluginhandler._part_environment import (
+    get_snapcraft_global_environment,
+)
+from snapcraft_legacy.project import Project
+from tests.legacy.fixture_setup import SnapcraftYaml
+
+
+def test_no_build_for_arch(tmp_path):
+    """build-for envvars should exist when the build-for arch can be determined."""
+    snapcraft_yaml = SnapcraftYaml(
+        tmp_path,
+        base="core20",
+        parts={"test-part": {"plugin": "nil"}},
+        architectures=[{"build-on": ["amd64", "arm64"], "run-on": "amd64"}],
+    )
+    snapcraft_yaml.write_snapcraft_yaml()
+    project = Project(snapcraft_yaml_file_path=snapcraft_yaml.snapcraft_yaml_file_path)
+
+    environment = get_snapcraft_global_environment(project)
+
+    assert environment["SNAPCRAFT_ARCH_BUILD_ON"] == project.arch_build_on
```

Comment:
This might be more obvious (ditto to other places,
```suggestion
    assert environment["SNAPCRAFT_ARCH_BUILD_ON"] == "amd64"
```
And so on, if you are thinking about the case of failure when running these tests on arm64, well, they will also fail on s390x. The platform.machine call needs to be mocked for that :-) Not a rush here